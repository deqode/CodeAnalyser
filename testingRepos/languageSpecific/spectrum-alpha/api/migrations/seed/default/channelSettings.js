const constants = require('./constants');
const { PAYMENTS_PRIVATE_CHANNEL_ID } = constants;

module.exports = [
  {
    id: 1,
    channelId: PAYMENTS_PRIVATE_CHANNEL_ID,
    joinSettings: {
      tokenJoinEnabled: true,
      token: 'abc',
    },
    slackSettings: {
      botLinks: {
        threadCreated: null,
      },
    },
  },
];
