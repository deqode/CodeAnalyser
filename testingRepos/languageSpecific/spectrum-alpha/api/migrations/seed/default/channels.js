// @flow
const constants = require('./constants');
const {
  DATE,
  SPECTRUM_COMMUNITY_ID,
  PAYMENTS_COMMUNITY_ID,
  DELETED_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
  SINGLE_CHANNEL_COMMUNITY_ID,
  SPECTRUM_GENERAL_CHANNEL_ID,
  SPECTRUM_PRIVATE_CHANNEL_ID,
  PAYMENTS_GENERAL_CHANNEL_ID,
  PAYMENTS_PRIVATE_CHANNEL_ID,
  PAYMENTS_FEATURES_CHANNEL_ID,
  SPECTRUM_ARCHIVED_CHANNEL_ID,
  SPECTRUM_DELETED_CHANNEL_ID,
  DELETED_COMMUNITY_DELETED_CHANNEL_ID,
  MODERATOR_CREATED_CHANNEL_ID,
  PRIVATE_GENERAL_CHANNEL_ID,
  SINGLE_CHANNEL_COMMUNITY_GENERAL_CHANNEL_ID,
} = constants;

module.exports = [
  {
    id: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General chatter',
    slug: 'general',
    isPrivate: false,
    isDefault: true,
    memberCount: 5,
  },

  {
    id: SPECTRUM_PRIVATE_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Private',
    description: 'Private chatter',
    slug: 'private',
    isPrivate: true,
    isDefault: false,
    memberCount: 5,
  },

  {
    id: PAYMENTS_GENERAL_CHANNEL_ID,
    communityId: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General chatter',
    slug: 'general',
    isPrivate: false,
    isDefault: true,
    memberCount: 5,
  },

  {
    id: PAYMENTS_PRIVATE_CHANNEL_ID,
    communityId: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Private',
    description: 'Private chatter',
    slug: 'private',
    isPrivate: true,
    isDefault: false,
    memberCount: 5,
  },

  {
    id: PAYMENTS_FEATURES_CHANNEL_ID,
    communityId: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Payments Features',
    description: 'Payments Features',
    slug: 'features',
    isPrivate: false,
    isDefault: false,
    memberCount: 5,
  },

  {
    id: SPECTRUM_ARCHIVED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Archived',
    description: 'Testing archiving',
    slug: 'archived',
    isPrivate: false,
    isDefault: true,
    archivedAt: new Date(DATE),
    memberCount: 3,
  },

  {
    id: SPECTRUM_DELETED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Deleted',
    description: 'Testing deleted channel',
    slug: 'deleted',
    isPrivate: false,
    isDefault: false,
    deletedAt: new Date(DATE),
    memberCount: 0,
  },

  {
    id: DELETED_COMMUNITY_DELETED_CHANNEL_ID,
    communityId: DELETED_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Deleted',
    description: 'Testing deleted channel',
    slug: 'deleted',
    isPrivate: false,
    isDefault: false,
    deletedAt: new Date(DATE),
    memberCount: 1,
  },

  {
    id: MODERATOR_CREATED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Moderator created',
    description: 'Moderator created channel',
    slug: 'moderator-created',
    isPrivate: false,
    isDefault: false,
    memberCount: 1,
  },

  {
    id: PRIVATE_GENERAL_CHANNEL_ID,
    communityId: PRIVATE_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General',
    slug: 'private-general',
    isPrivate: false,
    isDefault: false,
    memberCount: 1,
  },

  {
    id: SINGLE_CHANNEL_COMMUNITY_GENERAL_CHANNEL_ID,
    communityId: SINGLE_CHANNEL_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General',
    slug: 'general',
    isPrivate: false,
    isDefault: false,
    memberCount: 1,
  },
];
