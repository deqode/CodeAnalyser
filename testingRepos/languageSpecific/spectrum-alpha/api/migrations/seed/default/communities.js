// @flow
const constants = require('./constants');
const {
  DATE,
  SPECTRUM_COMMUNITY_ID,
  PAYMENTS_COMMUNITY_ID,
  DELETED_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
  SINGLE_CHANNEL_COMMUNITY_ID,
  PRIVATE_COMMUNITY_WITH_JOIN_TOKEN_ID,
} = constants;

module.exports = [
  {
    id: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: false,
    name: 'Spectrum',
    description: 'The future of communities',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'spectrum',
    memberCount: 4,
  },
  {
    id: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: false,
    name: 'Payments',
    description: 'Where payments are tested',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'payments',
    memberCount: 5,
  },
  {
    id: DELETED_COMMUNITY_ID,
    createdAt: new Date(DATE),
    deletedAt: new Date(DATE),
    isPrivate: false,
    name: 'Deleted',
    description: 'Things didnt work out',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'deleted',
    memberCount: 0,
  },
  {
    id: PRIVATE_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: true,
    name: 'Private community',
    description: 'Private community',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'private',
    memberCount: 1,
  },
  {
    id: SINGLE_CHANNEL_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: false,
    name: 'Single channel community',
    description: 'Single channel community',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'single',
    memberCount: 1,
  },
  {
    id: PRIVATE_COMMUNITY_WITH_JOIN_TOKEN_ID,
    createdAt: new Date(DATE),
    isPrivate: true,
    name: 'private community with join token',
    description: 'private community with join token',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'private-join',
    memberCount: 1,
  },
];
