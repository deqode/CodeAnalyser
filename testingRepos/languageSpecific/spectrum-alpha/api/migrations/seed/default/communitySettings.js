const constants = require('./constants');
const {
  PRIVATE_COMMUNITY_WITH_JOIN_TOKEN_ID,
  PAYMENTS_COMMUNITY_ID,
} = constants;

module.exports = [
  {
    id: 1,
    communityId: PRIVATE_COMMUNITY_WITH_JOIN_TOKEN_ID,
    brandedLogin: {
      isEnabled: false,
      message: null,
    },
    slackSettings: {
      connectedAt: null,
      connectedBy: null,
      teamName: null,
      teamId: null,
      scope: null,
      token: null,
      invitesSentAt: null,
      invitesMemberCount: null,
      invitesCustomMessage: null,
    },
    joinSettings: {
      tokenJoinEnabled: true,
      token: 'abc',
    },
  },
  {
    id: 2,
    communityId: PAYMENTS_COMMUNITY_ID,
    brandedLogin: {
      isEnabled: false,
      message: null,
    },
    slackSettings: {
      connectedAt: null,
      connectedBy: null,
      teamName: null,
      teamId: null,
      scope: null,
      token: null,
      invitesSentAt: null,
      invitesMemberCount: null,
      invitesCustomMessage: null,
    },
    joinSettings: {
      tokenJoinEnabled: true,
      token: 'abc',
    },
  },
];
