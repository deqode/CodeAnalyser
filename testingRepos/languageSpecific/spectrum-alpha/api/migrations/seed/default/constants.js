// @flow
const DATE = 1483225200000;

// users
const MAX_ID = '1';
const BRIAN_ID = '2';
const BRYN_ID = '3';
// this user is blocked in spectrum community
const BLOCKED_USER_ID = '4';
// this user is has never joined communities or channels
const QUIET_USER_ID = '5';
// this user was a previous member of spectrum community
const PREVIOUS_MEMBER_USER_ID = '6';
// this user is pending in all private channels
const PENDING_USER_ID = '7';
// this user is moderator in all channels, member in all communities
const CHANNEL_MODERATOR_USER_ID = '8';
// this user is moderator in all communities
const COMMUNITY_MODERATOR_USER_ID = '9';
// this user is only a member of one community, and that community only has
// one channel - use for testing the composer community+channel selection
const SINGLE_CHANNEL_COMMUNITY_USER_ID = '10';
const NEW_USER_ID = '11';

// communities
const SPECTRUM_COMMUNITY_ID = '1';
const PAYMENTS_COMMUNITY_ID = '2';
const DELETED_COMMUNITY_ID = '3';
const PRIVATE_COMMUNITY_ID = '4';
const SINGLE_CHANNEL_COMMUNITY_ID = '5';
const PRIVATE_COMMUNITY_WITH_JOIN_TOKEN_ID = '6';

// channels
const SPECTRUM_GENERAL_CHANNEL_ID = '1';
const SPECTRUM_PRIVATE_CHANNEL_ID = '2';
const PAYMENTS_GENERAL_CHANNEL_ID = '3';
const PAYMENTS_PRIVATE_CHANNEL_ID = '4';
const PAYMENTS_FEATURES_CHANNEL_ID = '10';
const SPECTRUM_ARCHIVED_CHANNEL_ID = '5';
const SPECTRUM_DELETED_CHANNEL_ID = '6';
const DELETED_COMMUNITY_DELETED_CHANNEL_ID = '7';
const MODERATOR_CREATED_CHANNEL_ID = '8';
const PRIVATE_GENERAL_CHANNEL_ID = '9';
const SINGLE_CHANNEL_COMMUNITY_GENERAL_CHANNEL_ID = '11';

module.exports = {
  DATE,
  MAX_ID,
  BRIAN_ID,
  BRYN_ID,
  BLOCKED_USER_ID,
  QUIET_USER_ID,
  PREVIOUS_MEMBER_USER_ID,
  PENDING_USER_ID,
  CHANNEL_MODERATOR_USER_ID,
  COMMUNITY_MODERATOR_USER_ID,
  SINGLE_CHANNEL_COMMUNITY_USER_ID,
  PRIVATE_COMMUNITY_WITH_JOIN_TOKEN_ID,
  NEW_USER_ID,
  SPECTRUM_COMMUNITY_ID,
  PAYMENTS_COMMUNITY_ID,
  DELETED_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
  SINGLE_CHANNEL_COMMUNITY_ID,
  SPECTRUM_GENERAL_CHANNEL_ID,
  SPECTRUM_PRIVATE_CHANNEL_ID,
  PAYMENTS_GENERAL_CHANNEL_ID,
  PAYMENTS_PRIVATE_CHANNEL_ID,
  PAYMENTS_FEATURES_CHANNEL_ID,
  SPECTRUM_ARCHIVED_CHANNEL_ID,
  SPECTRUM_DELETED_CHANNEL_ID,
  DELETED_COMMUNITY_DELETED_CHANNEL_ID,
  MODERATOR_CREATED_CHANNEL_ID,
  PRIVATE_GENERAL_CHANNEL_ID,
  SINGLE_CHANNEL_COMMUNITY_GENERAL_CHANNEL_ID,
};
