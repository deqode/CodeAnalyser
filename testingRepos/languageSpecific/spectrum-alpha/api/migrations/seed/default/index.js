// @flow
const constants = require('./constants');
const defaultUsers = require('./users');
const defaultCommunities = require('./communities');
const defaultChannels = require('./channels');
const defaultThreads = require('./threads');
const defaultUsersThreads = require('./usersThreads');
const defaultDirectMessageThreads = require('./directMessageThreads');
const defaultUsersDirectMessageThreads = require('./usersDirectMessageThreads');
const defaultUsersCommunities = require('./usersCommunities');
const defaultUsersChannels = require('./usersChannels');
const defaultUsersSettings = require('./usersSettings')();
const defaultMessages = require('./messages');
const defaultReactions = require('./reactions');
const defaultUsersNotifications = require('./usersNotifications');
const defaultNotifications = require('./notifications');
const defaultCommunitySettings = require('./communitySettings');
const defaultChannelSettings = require('./channelSettings');

module.exports = {
  constants,
  defaultUsers,
  defaultCommunities,
  defaultChannels,
  defaultThreads,
  defaultUsersThreads,
  defaultDirectMessageThreads,
  defaultUsersDirectMessageThreads,
  defaultUsersCommunities,
  defaultUsersChannels,
  defaultMessages,
  defaultUsersSettings,
  defaultNotifications,
  defaultUsersNotifications,
  defaultCommunitySettings,
  defaultChannelSettings,
  defaultReactions,
};
