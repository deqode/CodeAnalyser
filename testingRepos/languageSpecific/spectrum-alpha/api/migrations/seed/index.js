// @flow
const faker = require('faker');
const debug = require('debug')('api:migrations:seed');
const { v4: uuid } = require('uuid');
const {
  defaultCommunities,
  defaultUsers,
  defaultChannels,
  defaultThreads,
  defaultUsersThreads,
  defaultDirectMessageThreads,
  defaultUsersDirectMessageThreads,
  defaultUsersCommunities,
  defaultUsersChannels,
  defaultMessages,
} = require('./default/index');

const {
  randomAmount,
  generateUser,
  generateUsersSettings,
  generateCommunity,
  generateChannel,
  generateUsersCommunities,
  generateUsersChannels,
  generateThread,
  generateUsersThreads,
  generateDirectMessageThread,
  generateUsersDirectMessageThreads,
  generateMessage,
  generateReaction,
} = require('./generate');

const userAmount = faker.random.number(1000);
const users = [
  ...defaultUsers,
  ...randomAmount({ max: userAmount, min: 1 }, generateUser),
];

const usersSettings = [];
users.forEach(user => {
  usersSettings.push(generateUsersSettings(user.id));
});

debug('Generating communities...');
const communities = [
  ...defaultCommunities,
  ...randomAmount({ min: 10, max: 20 }, () => {
    return generateCommunity();
  }),
];

debug('Generating usersCommunities...');
let usersCommunities = [];
users.forEach(user => {
  communities.forEach(community => {
    usersCommunities.push(generateUsersCommunities(community.id, user.id));
  });
});

debug('Generating channels...');
let channels = defaultChannels;
communities.forEach(community => {
  if (community.deletedAt) return;
  randomAmount({ max: 10 }, () => {
    channels.push(generateChannel(community.id));
  });
});

debug('Generating usersChannels...');
let usersChannels = defaultUsersChannels;
const generatedUsersChannels = users.map(user => {
  return generateUsersChannels(channels, usersCommunities, user.id);
});
generatedUsersChannels.map(elem => {
  usersChannels.push(...elem);
});

debug('Generating threads...');
let threads = defaultThreads;
channels.forEach(channel => {
  const community = communities.find(
    community => community.id === channel.communityId
  );
  if (community.deletedAt) return;

  randomAmount({ max: 10 }, () => {
    const creator = faker.random.arrayElement(users);
    const thread = generateThread(channel.communityId, channel.id, creator.id);
    threads.push(thread);
  });
});

let usersThreads = defaultUsersThreads;
threads.forEach(thread => {
  const usersThread = generateUsersThreads(thread.id, thread.creatorId);
  usersThreads.push(usersThread);
});

debug('Generating direct message threads...');
let directMessageThreads = defaultDirectMessageThreads;
randomAmount({ max: 100 }, () => {
  directMessageThreads.push(generateDirectMessageThread());
});

debug('Generating usersDirectMessageThreads...');
let usersDirectMessageThreads = defaultUsersDirectMessageThreads;
directMessageThreads.forEach(thread => {
  const thread_users = randomAmount({ max: 5, min: 2 }, i => users[i]);

  thread_users.forEach(user => {
    usersDirectMessageThreads.push(
      generateUsersDirectMessageThreads(thread.id, user.id)
    );
  });
});

debug('Generating messages...');
let messages = defaultMessages;
threads.forEach(thread => {
  const channel = channels.find(channel => channel.id === thread.channelId);
  const threadMessages = [];
  randomAmount({ max: 10 }, () => {
    const sender = faker.random.arrayElement(users);
    const message = generateMessage(sender.id, thread.id, 'story');
    messages.push(message);
    threadMessages.push(message);
  });
});

debug('Generating direct messages...');
let direct_messages = [];
usersDirectMessageThreads.forEach(thread => {
  const threadMessages = [];
  const sender = thread.userId;
  randomAmount({ max: 100 }, () => {
    const message = generateMessage(
      sender,
      thread.threadId,
      'directMessageThread'
    );
    direct_messages.push(message);
    threadMessages.push(message);
  });
});

debug('Generating reactions...');
let reactions = [];
messages.map(message => {
  randomAmount({ max: 5 }, () => {
    const user = faker.random.arrayElement(users);
    reactions.push(generateReaction(user.id, message.id));
  });
});

debug('Connecting to db...');
// $FlowFixMe
const db = require('rethinkhaberdashery')({
  db: 'spectrum',
});

debug(
  `Inserting ${users.length} users,
  ${communities.length} communities, ${channels.length} channels, ${
    threads.length
  } threads, ${messages.length + direct_messages.length} messages, ${
    reactions.length
  } reactions, ${directMessageThreads.length} direct message threads, ${
    usersCommunities.length
  } usersCommunities objects, ${
    usersChannels.length
  } usersChannels objects, and ${
    usersDirectMessageThreads.length
  } usersDirectMessageThreads objects into the database... (this might take a while!)`
);
Promise.all([
  db
    .table('communities')
    .insert(communities)
    .run(),
  db
    .table('channels')
    .insert(channels)
    .run(),
  db
    .table('threads')
    .insert(threads)
    .run(),
  db
    .table('messages')
    .insert(messages)
    .run(),
  db
    .table('users')
    .insert(users)
    .run(),
  db
    .table('usersSettings')
    .insert(usersSettings)
    .run(),
  db
    .table('reactions')
    .insert(reactions)
    .run(),
  db
    .table('directMessageThreads')
    .insert(directMessageThreads)
    .run(),
  db
    .table('messages')
    .insert(direct_messages)
    .run(),
  db
    .table('usersCommunities')
    .insert(usersCommunities)
    .run(),
  db
    .table('usersChannels')
    .insert(usersChannels)
    .run(),
  db
    .table('usersDirectMessageThreads')
    .insert(usersDirectMessageThreads)
    .run(),
  db
    .table('usersThreads')
    .insert(usersThreads)
    .run(),
])
  .then(() => {
    debug('Finished seeding database! üéâ');
    process.exit();
  })
  .catch(err => {
    debug(
      'Encountered error while inserting data (see below), please run yarn run db:drop and yarn run db:migrate to restore tables to original condition, then run this script again.'
    );
    console.error('‚ùå Error in job:\n');
    console.error(err);
  });
