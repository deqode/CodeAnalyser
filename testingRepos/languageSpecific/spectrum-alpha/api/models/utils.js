// @flow
import { db } from 'shared/db';

export const NEW_DOCUMENTS = db
  .row('old_val')
  .eq(null)
  .and(db.not(db.row('new_val').eq(null)));

export const listenToNewDocumentsIn = (table: string, cb: Function) => {
  return (
    db
      .table(table)
      .changes({
        includeInitial: false,
      })
      // Filter to only include newly inserted messages in the changefeed
      .filter(NEW_DOCUMENTS)
      .run({ cursor: true })
      .then(cursor => {
        cursor.each((err, data) => {
          if (err) throw err;
          // Call the passed callback with the message directly
          cb(data.new_val);
        });
        return cursor;
      })
  );
};

export type Timeframe = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export const parseRange = (timeframe?: Timeframe) => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(row =>
      row
        .hasFields('lastSeen')
        .and(row('lastSeen').during(db.now().sub(current), db.now()))
    )
    .count()
    .default(0)
    .run();
};

// prettier-ignore
export const getGrowth = async (table: string, range: Timeframe, field: string, filter: ?mixed) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};

export const getCoreMetrics = () => {
  return db
    .table('coreMetrics')
    .orderBy(db.desc('date'))
    .limit(90)
    .orderBy('date')
    .run();
};
