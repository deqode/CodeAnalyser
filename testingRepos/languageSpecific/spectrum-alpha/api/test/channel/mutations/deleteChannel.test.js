//@flow
import { request } from '../../utils';
import db from 'shared/testing/db';
import data from 'shared/testing/data';
import { MAX_ID, BRYN_ID, DATE } from '../../../migrations/seed/default';

const generalChannelId = data.channels[0].id;
const owner = data.users.find(({ username }) => username === 'mxstbr');
const noPermissionUser = data.users.find(({ username }) => username === 'bryn');
const defaultChannelId = 'ce2b4488-4c75-47e0-8ebc-2539c';
const DEFAULT_CHANNELS = [
  {
    id: defaultChannelId,
    communityId: 'ce2b4488-4c75-47e0-8ebc-2539c1e6a191',
    createdAt: new Date(),
    name: 'Default',
    description: 'Default chatter',
    slug: 'default',
    isPrivate: false,
    isDefault: true,
  },
];
const DEFAULT_USERS_CHANNELS = [
  {
    createdAt: new Date(),
    userId: MAX_ID,
    channelId: defaultChannelId,
    isOwner: true,
    isModerator: false,
    isMember: true,
    isBlocked: false,
    isPending: false,
    receiveNotifications: true,
  },
  {
    createdAt: new Date(),
    userId: BRYN_ID,
    channelId: defaultChannelId,
    isOwner: false,
    isModerator: false,
    isMember: true,
    isBlocked: false,
    receiveNotifications: true,
  },
];
const DEFAULT_THREADS = [
  {
    createdAt: new Date(DATE),
    creatorId: MAX_ID,
    channelId: defaultChannelId,
    communityId: 'ce2b4488-4c75-47e0-8ebc-2539c1e6a191',
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'The first thread! ðŸŽ‰',
      body: '',
    },
    edits: [],
    modifiedAt: new Date(DATE),
    lastActive: new Date(DATE),
  },
  {
    createdAt: new Date(DATE + 1),
    creatorId: MAX_ID,
    channelId: defaultChannelId,
    communityId: 'ce2b4488-4c75-47e0-8ebc-2539c1e6a191',
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Another thread',
      body: '',
    },
    edits: [],
    modifiedAt: new Date(DATE + 1),
    lastActive: new Date(DATE + 1),
  },
  {
    createdAt: new Date(DATE + 2),
    creatorId: BRYN_ID,
    channelId: defaultChannelId,
    communityId: 'ce2b4488-4c75-47e0-8ebc-2539c1e6a191',
    isPublished: true,
    isLocked: false,
    type: 'DRAFTJS',
    content: {
      title: 'Yet another thread',
      body: '',
    },
    edits: [],
    modifiedAt: new Date(DATE + 2),
    lastActive: new Date(DATE + 2),
  },
];

const cleanTables = () =>
  Promise.all([
    db
      .table('channels')
      .delete()
      .run(),
    db
      .table('usersChannels')
      .delete()
      .run(),
    db
      .table('threads')
      .delete()
      .run(),
  ]);

const populateTables = () =>
  Promise.all([
    db
      .table('channels')
      .insert([...DEFAULT_CHANNELS, ...data.channels])
      .run(),
    db
      .table('usersChannels')
      .insert([...DEFAULT_USERS_CHANNELS, ...data.usersChannels])
      .run(),
    db
      .table('threads')
      .insert([...DEFAULT_THREADS, ...data.threads])
      .run(),
  ]);

// after each test just rest the database
beforeEach(() => cleanTables().then(() => populateTables()));
afterEach(() => cleanTables().then(() => populateTables()));

const variables = {
  channelId: defaultChannelId,
};

it('should delete a channel if user is owner', async () => {
  const query = /* GraphQL */ `
    mutation deleteChannel($channelId: ID!) {
      deleteChannel(channelId: $channelId)
    }
  `;

  const context = {
    user: owner,
  };

  expect.assertions(1);

  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
});

it('should not delete a channel if user is not owner', async () => {
  const query = /* GraphQL */ `
    mutation deleteChannel($channelId: ID!) {
      deleteChannel(channelId: $channelId)
    }
  `;

  const context = {
    user: noPermissionUser,
  };

  expect.assertions(1);

  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
});

it('should not delete a channel if user is not signed in', async () => {
  const query = /* GraphQL */ `
    mutation deleteChannel($channelId: ID!) {
      deleteChannel(channelId: $channelId)
    }
  `;

  expect.assertions(1);

  const result = await request(query, { variables });
  expect(result).toMatchSnapshot();
});

it('should not delete the general channel', async () => {
  const query = /* GraphQL */ `
    mutation deleteChannel($channelId: ID!) {
      deleteChannel(channelId: $channelId)
    }
  `;

  const context = {
    user: owner,
  };

  expect.assertions(1);

  const result = await request(query, {
    context,
    variables: {
      channelId: generalChannelId,
    },
  });
  expect(result).toMatchSnapshot();
});

it('should delete all threads in the deleted channel', async () => {
  const getThreadsInChannel = () =>
    db
      .table('threads')
      .filter({ channelId: defaultChannelId })
      .run();

  const query = /* GraphQL */ `
    mutation deleteChannel($channelId: ID!) {
      deleteChannel(channelId: $channelId)
    }
  `;

  const context = {
    user: owner,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
  const threads = await getThreadsInChannel();
  const filtered = threads.filter(t => t.deletedAt !== null).length;
  expect(threads).toHaveLength(filtered);
});
