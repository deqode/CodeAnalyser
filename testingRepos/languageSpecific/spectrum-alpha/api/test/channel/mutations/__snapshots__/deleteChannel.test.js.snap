// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should delete a channel if user is owner 1`] = `
Object {
  "data": Object {
    "deleteChannel": true,
  },
}
`;

exports[`should delete all threads in the deleted channel 1`] = `
Object {
  "data": Object {
    "deleteChannel": true,
  },
}
`;

exports[`should not delete a channel if user is not owner 1`] = `
Object {
  "data": Object {
    "deleteChannel": null,
  },
  "errors": Array [
    [GraphQLError: You don't have permission to make changes to this channel],
  ],
}
`;

exports[`should not delete a channel if user is not signed in 1`] = `
Object {
  "data": Object {
    "deleteChannel": null,
  },
  "errors": Array [
    [GraphQLError: You must be signed in to make changes to this channel.],
  ],
}
`;

exports[`should not delete the general channel 1`] = `
Object {
  "data": Object {
    "deleteChannel": null,
  },
  "errors": Array [
    [GraphQLError: The general channel can't be deleted],
  ],
}
`;
